### ISSUE TITLE
Signature of decorated function is not type checked by mypy

### ISSUE THREAD

**Issue by truenicoco** [aio-libs/async-lru#504]
The following has mypy  (expectedly) failing with `error: Argument 1 to "f1" has incompatible type "int"; expected "str"  [arg-type]`

```python
async def f1(x: str, y: int) -> str:
    return x + str(y)


async def main():
    await f1(1, 1)
```

But decorating `f1` with `@alru_cache` prevents mypy from detecting any issue.

**Comment by Dreamsorcerer** [comment]
Yes, it currently only types the return type correctly. It would need updating to use ParamSpec to type check the parameters.

**Comment by Dreamsorcerer** [comment]
I've started #508 for this, but it's not working for methods correctly, not sure what's gone wrong there. It also results in type errors for partial() as mypy/typeshed don't fully support it yet.

If anyone wants to try and make any fixes, that'd be great.

**Comment by rkjdid** [comment]
may be related https://youtrack.jetbrains.com/issue/PY-23067/Pycharm-not-picking-function-metadata-from-functools.wraps-with-methods#focus=Comments-27-6431674.0-0

I was having issues like this with (some) decorators in jetbrain's ide, the "preserved_signature" decorator on decorator fixes it
